src/qpx_mmc_css.cpp

```diff
@@ -0,0 +1,1909 @@
+
+  /* For all cycle length from 2 to 48 */
+  for (i = 2; i < 0x30; i++) {
+    /* Find the number of bytes that repeats in cycles. */
+    for (j = i + 1; j < 0x80 && (p_sec[0x7F - (j % i)] == p_sec[0x7F - j]);
+         j++) {
+      /* We have found j repeating bytes with a cycle length i. */
+      if (j > i_best_plen) {
+        i_best_plen = j;
+        i_best_p = i;
+      }
+    }
+  }
+
+  /* We need at most 10 plain text bytes?, so a make sure that we
+   * have at least 20 repeated bytes and that they have cycled at
+   * least one time.  */
+  if ((i_best_plen > 3) && (i_best_plen / i_best_p >= 2)) {
+    int res;
+
+    i_tries++;
+    memset(p_key, 0, DVD_KEY_SIZE);
+    res = css_RecoverTitleKey(
+        0, &p_sec[0x80], &p_sec[0x80 - (i_best_plen / i_best_p) * i_best_p],
+        &p_sec[0x54] /* key_seed */, p_key);
+    i_success += (res >= 0);
+#if 0
+        if( res >= 0 )
+        {
+            fprintf( stderr, "key is %02x:%02x:%02x:%02x:%02x ",
+                     p_key[0], p_key[1], p_key[2], p_key[3], p_key[4] );
+            fprintf( stderr, "at block %5d pattern len %3d period %3d %s\n",
+                     i_pos, i_best_plen, i_best_p, (res>=0?"y":"n") );
+        }
+#endif
+    return (res >= 0);
+  }
+
+  return 0;
+}
+
+/*****************************************************************************
+ * _dvdcss_title: crack or decrypt the current title key if needed
+ *****************************************************************************
+ * This function should only be called by dvdcss->pf_seek and should eventually
+ * not be external if possible.
+ *****************************************************************************/
+int css_title(drive_info *drive, int lba) {
+  dvd_title_t *p_title;
+  dvd_title_t *p_newtitle;
+  dvd_key_t p_title_key;
+  int i_ret = -1;
+#if (DVDCSS_KEY_CACHE > 0)
+  int i_fd, b_cache = 0;
+#endif
+
+  if (!drive->media.dvdcss.protection == 0x01) {
+    return 0;
+  }
+
+  /* Check if we've already cracked this key */
+  p_title = drive->media.dvdcss.p_titles;
+  while (p_title != NULL && p_title->p_next != NULL &&
+         p_title->p_next->i_startlb <= lba) {
+    p_title = p_title->p_next;
+  }
+
+  if (p_title != NULL && p_title->i_startlb == lba) {
+    /* We've already cracked this key, nothing to do */
+    memcpy(drive->media.dvdcss.TK, p_title->p_key, sizeof(dvd_key_t));
+    return 0;
+  }
+
+#if (DVDCSS_KEY_CACHE > 0)
+
+  /* Check whether the key is in our disk cache */
+  if (drive->media.dvdcss.psz_cachefile[0]) {
+    /* XXX: be careful, we use sprintf and not snprintf */
+    sprintf(drive->media.dvdcss.psz_block, "%.10x", lba);
+    i_fd = open(drive->media.dvdcss.psz_cachefile, O_RDONLY);
+    b_cache = 1;
+
+    if (i_fd >= 0) {
+      char psz_key[DVD_KEY_SIZE * 3];
+      unsigned int k0, k1, k2, k3, k4;
+
+      psz_key[DVD_KEY_SIZE * 3 - 1] = '\0';
+
+      if (read(i_fd, psz_key, DVD_KEY_SIZE * 3 - 1) == DVD_KEY_SIZE * 3 - 1 &&
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)